### PR TITLE
[docs] Add documentation for per query limits with config and headers

### PR DESCRIPTION
**Pull request by robskillington**
**What this PR does / why we need it**:

Adds missing documentation for per query limits by configuration and headers.

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: docs/operational_guide/resource_limits.md [hunk 1]
```diff
@@ -4,16 +4,93 @@ This operational guide provides an overview of how to set resource limits on
...
 performance of M3 in a production environment.
 
+## M3 Query and M3 Coordinator
+
+### Deployment
+
+Protecting queries impacting your ingestion of metrics for metrics workloads 
+can first and foremost be done by deploying M3 Query and M3 Coordinator 
+independently. That is, for writes to M3 use a dedicated deployment of 
+M3 Coordinator instances, and then for queries to m3 use a dedicated deployment
```
Comment (rallen090) [inline docs/operational_guide/resource_limits.md:68]: nit: capitalize "M3"

File: docs/operational_guide/resource_limits.md [hunk 2]
```diff
@@ -4,16 +4,93 @@ This operational guide provides an overview of how to set resource limits on
...
+M3 Coordinator instances, and then for queries to m3 use a dedicated deployment 
+of M3 Query instances.
+
+This ensures when M3 Query instances become busy and are starved of resources 
+serving an unexpected query load, they will not interrupt the flow of metrics
+being ingested to M3.
+
+### Configuring limits
+
+To protect individual queries using too many resources, you can specify some
```
Comment (rallen090) [inline docs/operational_guide/resource_limits.md:77]: nit: "To protect *against*..."

File: docs/operational_guide/resource_limits.md [hunk 3]
```diff
@@ -54,3 +57,77 @@ limits:
...
```
Comment (robskillington) [inline docs/operational_guide/resource_limits.md:77]:
```suggestion
To protect against individual queries using too many resources, you can specify some
```

File: docs/operational_guide/resource_limits.md [hunk 4]
```diff
@@ -54,3 +57,77 @@ limits:
...
 ```
+
+## M3 Query and M3 Coordinator
+
+### Deployment
+
+Protecting queries impacting your ingestion of metrics for metrics workloads 
+can first and foremost be done by deploying M3 Query and M3 Coordinator 
+independently. That is, for writes to M3 use a dedicated deployment of 
+M3 Coordinator instances, and then for queries to m3 use a dedicated deployment
```
Comment (robskillington) [inline docs/operational_guide/resource_limits.md:68]:
```suggestion
M3 Coordinator instances, and then for queries to M3 use a dedicated deployment 
```

File: docs/operational_guide/resource_limits.md [hunk 5]
```diff
@@ -4,16 +4,93 @@ This operational guide provides an overview of how to set resource limits on
...
+    # If set limits the number of time series returned for any given 
+    # individual storage node per query, before returning result to query 
+    # service.
+    maxFetchedSeries: 0
+
+    # If set limits the number of index documents matched for any given 
+    # individual storage node per query, before returning result to query 
+    # service.
+    # This equates to the number of time series * number of blocks, so for 
+    # 100 time series matching 4 hours of data for a namespace using a 2 hour
```
Comment (rallen090) [inline docs/operational_guide/resource_limits.md:106]: Hm wouldn't this be `100 series * 4 hours / 2 hour block size` == `200 docs`?

File: docs/operational_guide/resource_limits.md [hunk 6]
```diff
@@ -4,16 +4,93 @@ This operational guide provides an overview of how to set resource limits on
...
+    # 100 time series matching 4 hours of data for a namespace using a 2 hour 
+    # block size, that would result in matching 400 index documents.
+    maxFetchedDocs: 0
+
+    # If true this results in causing a query error if the query exceeds 
+    # the series limit for any given individual storage node per query.
+    requireExhaustive: false
+
+    # If set this limits the max number of datapoints allowed to be used by a
+    # given query, this is applied at the query service after the result has
```
Comment (rallen090) [inline docs/operational_guide/resource_limits.md:115]: nit: should "This is applied..." be a new sentence?

File: docs/operational_guide/resource_limits.md [hunk 7]
```diff
@@ -54,3 +57,77 @@ limits:
...
+    # If set limits the number of index documents matched for any given 
+    # individual storage node per query, before returning result to query 
+    # service.
+    # This equates to the number of time series * number of blocks, so for 
+    # 100 time series matching 4 hours of data for a namespace using a 2 hour 
+    # block size, that would result in matching 400 index documents.
+    maxFetchedDocs: 0
+
+    # If true this results in causing a query error if the query exceeds 
+    # the series limit for any given individual storage node per query.
```
Comment (robskillington) [inline docs/operational_guide/resource_limits.md:111]:
```suggestion
    # the series or blocks limit for any given individual storage node per query.
```

File: docs/operational_guide/resource_limits.md [hunk 8]
```diff
@@ -4,16 +4,93 @@ This operational guide provides an overview of how to set resource limits on
...
+  global:
+    # If set this limits the max number of datapoints allowed to be used by all
+    # queries at any point in time, this is applied at the query service after 
+    # the result has been returned by a storage node.
+    maxFetchedDatapoints: 0
+```
+
+### Headers
+
+The following headers can also be used to override configured limits on a per request basis (to allow for different limits dependendent on caller):
```
Comment (rallen090) [inline docs/operational_guide/resource_limits.md:129]: nit: "dependent"

File: docs/operational_guide/resource_limits.md (hunk 5 above)
Comment (robskillington) [reply docs/operational_guide/resource_limits.md:106]: Ah yes, heh, bad maths. Thank you.

File: docs/operational_guide/resource_limits.md [hunk 9]
```diff
@@ -54,3 +57,77 @@ limits:
...
+    # individual storage node per query, before returning result to query 
+    # service.
+    maxFetchedSeries: 0
+
+    # If set limits the number of index documents matched for any given 
+    # individual storage node per query, before returning result to query 
+    # service.
+    # This equates to the number of time series * number of blocks, so for 
+    # 100 time series matching 4 hours of data for a namespace using a 2 hour 
+    # block size, that would result in matching 400 index documents.
```
Comment (robskillington) [inline docs/operational_guide/resource_limits.md:107]:
```suggestion
    # block size, that would result in matching 200 index documents.
```

File: docs/operational_guide/resource_limits.md [hunk 10]
```diff
@@ -54,3 +57,77 @@ limits:
...
+    maxFetchedDocs: 0
+
+    # If true this results in causing a query error if the query exceeds 
+    # the series or blocks limit for any given individual storage node per query.
+    requireExhaustive: false
+
+    # If set this limits the max number of datapoints allowed to be used by a
+    # given query, this is applied at the query service after the result has
```
Comment (robskillington) [inline docs/operational_guide/resource_limits.md:115]:
```suggestion
    # given query. This is applied at the query service after the result has 
```

File: docs/operational_guide/resource_limits.md [hunk 11]
```diff
@@ -54,3 +57,77 @@ limits:
...
+  global:
+    # If set this limits the max number of datapoints allowed to be used by all
+    # queries at any point in time, this is applied at the query service after 
+    # the result has been returned by a storage node.
+    maxFetchedDatapoints: 0
+```
+
+### Headers
+
+The following headers can also be used to override configured limits on a per request basis (to allow for different limits dependendent on caller):
```
Comment (robskillington) [inline docs/operational_guide/resource_limits.md:129]:
```suggestion
The following headers can also be used to override configured limits on a per request basis (to allow for different limits dependent on caller):
```

File: src/cmd/services/m3query/config/config.go [hunk 12]
```diff
@@ -293,6 +294,24 @@ func (l *GlobalLimitsConfiguration) AsLimitManagerOptions() cost.LimitManagerOpt
...
+	// individual storage node per query, before returning result to query 
+	// service.
+	MaxFetchedSeries int `yaml:"maxFetchedSeries"`
+
+	// MaxFetchedDocs limits the number of index documents matched for any given 
+	// individual storage node per query, before returning result to query 
+	// service.
+	MaxFetchedDocs int `yaml:"maxFetchedDocs"`
+
+	// RequireExhaustive results in an error if the query exceeds the series limit.
```
Comment (rallen090) [inline src/cmd/services/m3query/config/config.go:307]: nit: "...  if the query exceeds the series limit" should probably say "any limit"

Comment (robskillington) [inline src/cmd/services/m3query/config/config.go:307]:
```suggestion
	// RequireExhaustive results in an error if the query exceeds any limit.
```

**Pull request merged**